### PR TITLE
Rails Associations Project: Private Events

### PR DESCRIPTION
**Pull request by alvarorf**
In this pull request, I attach the project Private Events for the Rails Curriculum. 
The purpose of this project was to implement associations in rails, in this case via the Private Events project. Therefore, users can create an account, login and logout, and only authorized users can create and attend (via an attendaces model/view/controller's combo) an event.
Thank you for your suggestions and any errors that you may find.

### CODE REVIEW

File: app/controllers/events_controller.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  before_action :set_event, only: %i[show edit update destroy]
+  before_action :require_user, except: %i[show index]
+  before_action :require_same_user, only: %i[edit update destroy]
+
+  # GET /events
+  # GET /events.json
+  def index
+    @events = Event.all
+    @future_events = @events.upcoming
+    @past_events = @events.past
```
Comment (elmejdki) [inline app/controllers/events_controller.rb:11]: no need for the first line, you could do just:

```ruby

     @future_events = Event.upcoming
     @past_events = Event.past

```

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,77 @@
+class UsersController < ApplicationController
+  before_action :set_user, only: %i[show edit update destroy]
+  before_action :require_login, only: %i[edit update]
+  before_action :require_same_user, only: %i[edit update destroy]
+
+  def show
+    if logged_in?
+      @future_events = current_user.events.upcoming
+      @past_events = current_user.events.past
```
Comment (elmejdki) [inline app/controllers/users_controller.rb:9]: * [ ] lists events created by user, i.e. `@created_events = current_user.events` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/current_user_show_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [ ] lists events previoulsy attendented by user **using event model scope**, i.e. `@past_events = current_user.attended_events.previous` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_events_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))
* [ ] lists upcoming events previously attended by user **using event model scope**, i.e. `@upcoming_events = current_user.attended_events.upcoming`

File: app/views/devise/mailer/email_changed.html.erb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+<p>Hello <%= @email %>!</p>
+
+<% if @resource.try(:unconfirmed_email?) %>
+  <p>We're contacting you to notify you that your email is being changed to <%= @resource.unconfirmed_email %>.</p>
+<% else %>
+  <p>We're contacting you to notify you that your email has been changed to <%= @resource.email %>.</p>
+<% end %>
```
Comment (elmejdki) [inline app/views/devise/mailer/email_changed.html.erb:7]: Please remove all the logic from your reviews and use view helpers to add logic into the code.

File: app/views/devise/passwords/edit.html.erb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+<%= form_for(resource, as: resource_name, url: password_path(resource_name), html: { method: :put }) do |f| %>
+  <%= render "devise/shared/error_messages", resource: resource %>
+  <%= f.hidden_field :reset_password_token %>
+
+  <div class="field">
+    <%= f.label :password, "New password" %><br />
+    <% if @minimum_password_length %>
+      <em>(<%= @minimum_password_length %> characters minimum)</em><br />
+    <% end %>
```
Comment (elmejdki) [inline app/views/devise/passwords/edit.html.erb:11]: Please remove all the logic from your reviews and use view helpers to add logic into the code.

File: app/views/devise/registrations/new.html.erb [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+  <div class="field">
+    <%= f.label :email %><br />
+    <%= f.email_field :email, autofocus: true, autocomplete: "email" %>
+  </div>
+
+  <div class="field">
+    <%= f.label :password %>
+    <% if @minimum_password_length %>
+    <em>(<%= @minimum_password_length %> characters minimum)</em>
+    <% end %><br />
```
Comment (elmejdki) [inline app/views/devise/registrations/new.html.erb:15]: Please remove all the logic from your reviews and use view helpers to add logic into the code.

File: app/views/devise/shared/_links.html.erb [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+<%- if devise_mapping.lockable? && resource_class.unlock_strategy_enabled?(:email) && controller_name != 'unlocks' %>
+  <%= link_to "Didn't receive unlock instructions?", new_unlock_path(resource_name) %><br />
+<% end %>
+
+<%- if devise_mapping.omniauthable? %>
+  <%- resource_class.omniauth_providers.each do |provider| %>
+    <%= link_to "Sign in with #{OmniAuth::Utils.camelize(provider)}", omniauth_authorize_path(resource_name, provider) %><br />
+  <% end %>
+<% end %>
```
Comment (elmejdki) [inline app/views/devise/shared/_links.html.erb:25]: Please remove all the logic from your reviews and use view helpers to add logic into the code.

**Review by elmejdki — CHANGES_REQUESTED** [review]
# status: require changes :recycle: 

Great job @alvarorf I really liked your project :clap: :clap: :clap: :clap:, even though you still have some changes :hammer: :wrench: :toolbox: to make to get it approved :checkered_flag:.

Please check the comments below :point_down: :point_down: :point_down: :point_down: :point_down: 

- [ ] Your users show page is super confusing, it should only display events the user has created, and events that the user is attending to separated in past and upcoming.

- [ ] please fix the header on the top it's super hard to follow the header where it's gonna land to pick the next link to click in.

- [ ]  Please remove all the logic from your reviews and use view helpers to add logic into the code.

- [ ] Please add Unit tests, and integration tests to your project.

* [x] User model: `has_many :events, foreign_key: :creator_id` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

* [ ] Event model `belongs_to :creator, class_name: "User"` ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/associations/user_assoc_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/associations))

- [ ] In your schema for events table you have user_id and creator_id, please pick one I would recommend creator_id, and don't forget to add an index for it.

### Please make all the required changes before you submit another code review.
## Happy Coding :tophat: 
_If you have any questions or you feel stuck don't hesitate to contact me on slack I will be happy to help._

**Review by javitocor — CHANGES_REQUESTED** [review]
# STATUS CHANGES REQUIRED 🛑 

Hello 👋 
A great effort implementing the changes requested by the previous TSE! 👍 
However, we still to make some changes in order to get the project approved 💪 
Kindly check my comments below 👇  implement the changes and request a new code review once done ⏩ 
Good luck! 😺 
## Mandatory changes
- Add unit tests to all the models, validations, associations and scopes
- Add integration tests to all the features of your app
- Remove test folder since you have your testing placed in spec folder
- Remove commented-out code in your controllers
- Remove json response since you are not using it in your controllers
- If you are not using actions, kindly remove it from your controller and routes
- In your schema for events table you have user_id and creator_id, please pick one I would recommend creator_id, and don't forget to add an index for it.
- And modify your models accordingly

**Review by Temmarie — CHANGES_REQUESTED** [review]
## Changes Required! :recycle:

### Hi @alvarorf  ! 👋 👋 

### You have done well so far :clap: However, there are changes required for you to make for this project to be approved. Kindly work on them and submit them for another code review.

**Recommended** 
- [ ] Add negative cases to your tests.
- [ ] Some of your tests are failing, kindly fix them.
**See here** 👇 
Screenshot (173)


__Kindly make the necessary changes and submit another code review request.__
_Feel free to reach me on slack for any inquiries about this review._ :wink:
__Grace Tamara Ekunola __
[GitHub](https://github.com/temmarie) | [LinkedIn](https://www.linkedin.com/in/ekunola-grace/) | [Twitter](https://twitter.com/TemmarieW)

**Review by PhillipUg — APPROVED** [review]
# Status: Project Approved! 🥇 
Congratulations on implementing all requirements and getting this project approved. :+1:

You can merge this branch into the development branch when you're ready and go on to your next project.
:bulb:  Do not forget to submit your project completion form after merging this branch.

All the best team,

For any assistance, you can contact me on slack as **Phillip Musiime**
